tests/system/src/test/java/test/com/sun/javafx/tk/quantum/SystemMenuBarTest.java

```diff
@@ -0,0 +1,351 @@
+                e.printStackTrace();
+                failed.set(true);
+                removeMenuLatch.countDown();
+            });
+            createRemoveMenuStage();
+        });
+        removeMenuLatch.await();
+        assertFalse(failed.get());
+    }
+
+    public void createRemoveMenuStage() {
+        Stage stage = new Stage();
+        VBox root = new VBox();
+
+        final MenuBar menuBar = new MenuBar();
+
+        Menu mainMenu = new Menu("MainMenu");
+        Menu subMenu = new Menu("SubMenu");
+        subMenu.getItems().addAll(new MenuItem("submenuitem1"), new MenuItem("submenuitem2"));
+        mainMenu.getItems().add(subMenu);
+        menuBar.getMenus().add(mainMenu);
+
+        menuBar.setUseSystemMenuBar(true);
+        root.getChildren().add(menuBar);
+
+        Scene scene = new Scene(root);
+        stage.setScene(scene);
+        stage.show();
+
+        Platform.runLater(() -> {
+            mainMenu.getItems().clear();
+            mainMenu.getItems().add(subMenu);
+            subMenu.getItems().addAll(new MenuItem("new item 1"), new MenuItem("new item 2"));
+            removeMenuLatch.countDown();
+        });
+    }
+
+    @Test
+    public void testJDK8309935() {
+        MenuBar menuBar = new MenuBar();
+        AtomicReference<Throwable> throwableRef = new AtomicReference<>();
+        Util.runAndWait(() -> {
+            Thread.currentThread().setUncaughtExceptionHandler((t, e) -> {
+                e.printStackTrace();
+                throwableRef.set(e);
+            });
+            menuBar.setUseSystemMenuBar(true);
+            Menu menu1 = new Menu("menu 1");
+            menu1.getItems().add(new MenuItem("item 1"));
+            menu1.getItems().add(new MenuItem("item 2"));
+            menuBar.getMenus().add(menu1);
+            Menu menu2 = new Menu(" menu 2");
+            menu2.getItems().add(new MenuItem("item 1"));
+            menu2.getItems().add(new MenuItem("item 2"));
+            menu2.getItems().add(new SeparatorMenuItem());
+            menuBar.getMenus().add(menu2);
+            Menu test1 = new Menu("test 1");
+            test1.getItems().add(new MenuItem("item 1"));
+            test1.getItems().add(new MenuItem("item 2"));
+            Menu test2 = new Menu("test 2");
+            test2.getItems().add(new MenuItem("item 1"));
+            test2.getItems().add(new MenuItem("item 2"));
+            menu2.addEventFilter(Menu.ON_SHOWING, e -> {
+                menu2.getItems().removeIf(o -> Objects.equals(o.getText(), test1.getText()));
+                menu2.getItems().add(test1);
+                menu2.getItems().removeIf(o -> Objects.equals(o.getText(), test2.getText()));
+                menu2.getItems().add(test2);
+            });
+            BorderPane root = new BorderPane();
+            root.setTop(menuBar);
+            Stage stage = new Stage();
+            stage.setScene(new Scene(root));
+            stage.show();
```

Comment:
I think, a CountDownLatch should be added to make sure that stage is shown before proceeding.

Reply:
I don't completely understand this. The stage.show() is a blocking call, and the runnable that it might schedule will be executed before other runnables (part of the processing in the rest of the method) are scheduled.
I added a CountDownLatch at the end of the processing, so that we allow for the processing to run into exceptions before we finalize the test.
Does that make sense?

Reply:
One correction: `Stage.show` is not a blocking call. I haven't looked at it closely, but perhaps what @arapte meant was to suggest that the test wait on a latch that is triggered when the Stage::showing event is delivered? Many of our headful tests do that, but I don't know whether that would be applicable here or not.

Reply:
Right, it wasn't correct to call it a "blocking call". However, before Stage.show() returns, the contents of the stage are created and their layoutChildren method is invoked (before the WINDOW_SHOWN is fired). 
All Runnables that are scheduled during this method will be executed before the content in the next Util.runAndWait() is executed, so I'm not sure what we should be waiting for?